### PR TITLE
Add alternative optional service, control, packet & button uuids

### PR DESCRIPTION
**Pull request by av-eric**
Added support for devices using non-standard UUIDS for DFU service and control, packet and button characteristics.

### CODE REVIEW

File: src/secure-dfu.ts [hunk 1]
```diff
@@ -212,15 +212,16 @@ export class SecureDfu extends EventDispatcher {
         });
     }
 
-    private gattConnect(device: BluetoothDevice): Promise<Array<BluetoothRemoteGATTCharacteristic>> {
+    private gattConnect(device: BluetoothDevice, serviceUUID?: number|string): Promise<Array<BluetoothRemoteGATTCharacteristic>> {
```
Comment (thegecko) [inline src/secure-dfu.ts]: nit: please use spaces between union types - `number | string`

File: src/secure-dfu.ts [hunk 2]
```diff
@@ -212,15 +212,16 @@ export class SecureDfu extends EventDispatcher {
...
+        serviceUUID = serviceUUID || SecureDfu.SERVICE_UUID;
```
Comment (thegecko) [inline src/secure-dfu.ts]: Prefer a default value on the method signature:

```typescript
private gattConnect(device: BluetoothDevice, serviceUUID: number | string = SecureDfu.SERVICE_UUID): Promise...

File: src/secure-dfu.ts [hunk 3]
```diff
@@ -393,15 +394,27 @@ export class SecureDfu extends EventDispatcher {
      * Scans for a device to update
      * @param buttonLess Scans for all devices and will automatically call `setDfuMode`
      * @param filters Alternative filters to use when scanning
+     * @param uuids Optional alternative uuids for service, control, packet or button
```
Comment (thegecko) [inline src/secure-dfu.ts:410]: Please extract uuids into an interface

File: src/secure-dfu.ts [hunk 4]
```diff
@@ -393,15 +394,27 @@ export class SecureDfu extends EventDispatcher {
...
      * @returns Promise containing the device
      */
-    public requestDevice(buttonLess: boolean, filters: Array<BluetoothLEScanFilterInit>): Promise<BluetoothDevice> {
+    public requestDevice(
+      buttonLess: boolean,
+      filters: Array<BluetoothLEScanFilterInit>,
+      uuids?: {
```
Comment (thegecko) [inline src/secure-dfu.ts]: Please use an interface to describe this

File: src/secure-dfu.ts [hunk 5]
```diff
@@ -393,15 +394,27 @@ export class SecureDfu extends EventDispatcher {
      * Scans for a device to update
      * @param buttonLess Scans for all devices and will automatically call `setDfuMode`
      * @param filters Alternative filters to use when scanning
+     * @param uuids Optional alternative uuids for service, control, packet or button
      * @returns Promise containing the device
      */
-    public requestDevice(buttonLess: boolean, filters: Array<BluetoothLEScanFilterInit>): Promise<BluetoothDevice> {
+    public requestDevice(
+      buttonLess: boolean,
```
Comment (thegecko) [inline src/secure-dfu.ts]: nit: 4 space indent to match rest of file

File: src/secure-dfu.ts [hunk 6]
```diff
@@ -419,27 +432,33 @@ export class SecureDfu extends EventDispatcher {
     /**
      * Sets the DFU mode of a device, preparing it for update
      * @param device The device to switch mode
+     * @param uuids Optional alternative uuids for control, packet or button
      * @returns Promise containing the device if it is still on a valid state
      */
-    public setDfuMode(device: BluetoothDevice): Promise<BluetoothDevice> {
-        return this.gattConnect(device)
+    public setDfuMode(device: BluetoothDevice, uuids?: { service?: number|string, control?: number|string, packet?: number|string, button?: number|string}): Promise<BluetoothDevice> {
```
Comment (thegecko) [inline src/secure-dfu.ts]: Please extract uuids as interface

File: src/secure-dfu.ts [hunk 7]
```diff
@@ -419,27 +432,33 @@ export class SecureDfu extends EventDispatcher {
...
+        uuids = uuids || {};
+        const serviceId = uuids.service || SecureDfu.SERVICE_UUID;
```
Comment (thegecko) [inline src/secure-dfu.ts]: used?

File: src/secure-dfu.ts [hunk 8]
```diff
@@ -419,27 +432,33 @@ export class SecureDfu extends EventDispatcher {
...
+        const controlId = uuids.control || CONTROL_UUID;
+        const packetId = uuids.packet || PACKET_UUID;
+        const buttonId = uuids.button || BUTTON_UUID;
+        return this.gattConnect(device, serviceId)
```
Comment (thegecko) [inline src/secure-dfu.ts]: nit: newline

**Review by thegecko — CHANGES_REQUESTED** [review]

**Comment by thegecko** [comment]
Thanks @av-eric, I've added some feedback.

File: src/secure-dfu.ts (hunk 6 above)
Comment (av-eric) [reply src/secure-dfu.ts]: Do you have a preference for the name of such interface?

Comment (thegecko) [reply src/secure-dfu.ts]: hmm `Uuids` :) or perhaps `UuidOptions`

Comment (av-eric) [reply src/secure-dfu.ts]: I like UuidOptions. I'll go with that.

File: src/secure-dfu.ts [hunk 9]
```diff
@@ -435,30 +435,25 @@ export class SecureDfu extends EventDispatcher {
...
      * @returns Promise containing the device if it is still on a valid state
      */
-    public setDfuMode(device: BluetoothDevice, uuids?: { service?: number|string, control?: number|string, packet?: number|string, button?: number|string}): Promise<BluetoothDevice> {
-        uuids = uuids || {};
-        const serviceId = uuids.service || SecureDfu.SERVICE_UUID;
-        const controlId = uuids.control || CONTROL_UUID;
-        const packetId = uuids.packet || PACKET_UUID;
-        const buttonId = uuids.button || BUTTON_UUID;
-        return this.gattConnect(device, serviceId)
+    public setDfuMode(device: BluetoothDevice, uuids: UuidOptions = this.DEFAULT_UUIDS): Promise<BluetoothDevice> {
```
Comment (thegecko) [inline src/secure-dfu.ts:448]: I like this approach, but it won't fill in the blanks if you only supply a few of the uuids.

You could make them all mandatory in the interface so it's an 'all or nothing' approach or do:

```typescript
public setDfuMode(device: BluetoothDevice, uuids: UuidOptions = this.DEFAULT_UUIDS): Promise<BluetoothDevice> {
    uuids = {
        ...this.DEFAULT_UUIDS,
        ...uuids
    };
    ...
```

File: src/uuid-options.ts [hunk 10]
```diff
@@ -0,0 +1,6 @@
+export interface UuidOptions {
```
Comment (thegecko) [inline src/uuid-options.ts]: Doesn't really need it's own file, you could keep it at the top of `secure-dfu.ts`?

File: src/secure-dfu.ts [hunk 11]
```diff
@@ -212,8 +220,7 @@ export class SecureDfu extends EventDispatcher {
         });
     }
 
-    private gattConnect(device: BluetoothDevice, serviceUUID?: number|string): Promise<Array<BluetoothRemoteGATTCharacteristic>> {
-        serviceUUID = serviceUUID || SecureDfu.SERVICE_UUID;
+    private gattConnect(device: BluetoothDevice, serviceUUID: number | string = SecureDfu.SERVICE_UUID): Promise<Array<BluetoothRemoteGATTCharacteristic>> {
```
Comment (thegecko) [inline src/secure-dfu.ts:229]: :+1:

File: src/secure-dfu.ts [hunk 12]
```diff
@@ -398,23 +405,16 @@ export class SecureDfu extends EventDispatcher {
...
-      filters: Array<BluetoothLEScanFilterInit>,
-      uuids?: {
-          service?: number|string,
-          button?: number|string,
-          control?: number|string,
-          packet?: number|string
-      }
+        buttonLess: boolean,
+        filters: Array<BluetoothLEScanFilterInit>,
+        uuids: UuidOptions = this.DEFAULT_UUIDS
```
Comment (thegecko) [inline src/secure-dfu.ts:416]: This may not cover situations where only a few uuids are passed (see below)

**Review by thegecko — CHANGES_REQUESTED** [review]

**Comment by thegecko** [comment]
Thanks for your additions, @av-eric

**Comment by av-eric** [comment]
Hey @thegecko! Any plans for a new release including this changes? We could really use that since adding a dependency to a branch is braking our build(with a './dist' not found error).

Thanks!

**Comment by thegecko** [comment]
It looks good to go, but still need to give it some more testing

**Pull request merged**